android/src/main/java/me/furtado/smsretriever/SmsBroadcastReceiver.java

```diff
@@ -6,79 +6,69 @@
 import android.os.Bundle;
 import android.support.annotation.NonNull;
 
-import com.facebook.react.bridge.ReactApplicationContext;
-import com.facebook.react.bridge.WritableNativeMap;
-import com.facebook.react.modules.core.DeviceEventManagerModule;
 import com.google.android.gms.auth.api.phone.SmsRetriever;
 import com.google.android.gms.common.api.CommonStatusCodes;
 import com.google.android.gms.common.api.Status;
 
 public final class SmsBroadcastReceiver extends BroadcastReceiver {
 
-    private static final String SMS_EVENT = "me.furtado.smsretriever:SmsEvent";
+    private SmsReceiveListener mSmsListener;
 
-    private static final String EXTRAS_KEY = "extras";
-    private static final String MESSAGE_KEY = "message";
-    private static final String STATUS_KEY = "status";
-    private static final String TIMEOUT_KEY = "timeout";
-
-    private static final String EXTRAS_NULL_ERROR_MESSAGE = "Extras is null.";
-    private static final String STATUS_NULL_ERROR_MESSAGE = "Status is null.";
-    private static final String TIMEOUT_ERROR_MESSAGE = "Timeout error.";
-
-    private ReactApplicationContext mContext;
-
-    public SmsBroadcastReceiver(final ReactApplicationContext context) {
-        mContext = context;
+    public void setSmsListener(SmsReceiveListener smsListener) {
+        mSmsListener = smsListener;
     }
 
     @Override
     public void onReceive(final Context context, final Intent intent) {
-        if (SmsRetriever.SMS_RETRIEVED_ACTION.equals(intent.getAction())) {
-            final Bundle extras = intent.getExtras();
-            if (extras == null) {
-                emitJSEvent(EXTRAS_KEY, EXTRAS_NULL_ERROR_MESSAGE);
-                return;
-            }
-
-            final Status status = (Status) extras.get(SmsRetriever.EXTRA_STATUS);
-            if (status == null) {
-                emitJSEvent(STATUS_KEY, STATUS_NULL_ERROR_MESSAGE);
-                return;
-            }
-
-            switch (status.getStatusCode()) {
-                case CommonStatusCodes.SUCCESS: {
-                    final String message = (String) extras.get(SmsRetriever.EXTRA_SMS_MESSAGE);
-                    emitJSEvent(MESSAGE_KEY, message);
-                    break;
-                }
-
-                case CommonStatusCodes.TIMEOUT: {
-                    emitJSEvent(TIMEOUT_KEY, TIMEOUT_ERROR_MESSAGE);
-                    break;
-                }
-            }
+        if (!SmsRetriever.SMS_RETRIEVED_ACTION.equals(intent.getAction())) {
+            return;
         }
-    }
-
-    //region - Privates
 
-    private void emitJSEvent(@NonNull final String key, final String message) {
-        if (mContext == null) {
+        if (mSmsListener == null) {
```

Comment:
Why even allow `mSmsListener` to be null? Why not throw an exception in `setSmsListener`?

Reply:
It would be null if someone forgets to set the mSmsListener before onReceive gets called.
Though maybe an exception would make sense. I'm not super familiar with 'standard' android error handling tactics. What happens if a broadcast receiver like this throws an exception?